homeassistant/components/bluecurrent/__init__.py

```diff
@@ -0,0 +1,244 @@
+    """Set up Blue Current as a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    client = Client()
+    api_token = config_entry.data[CONF_API_TOKEN]
+    connector = Connector(hass, config_entry, client)
+    try:
+        await connector.connect(api_token)
+    except (BlueCurrentException) as err:
+        if isinstance(err, InvalidApiToken):
+            raise ConfigEntryAuthFailed("Invalid api token.") from err
+        raise ConfigEntryNotReady from err
+
+    hass.loop.create_task(connector.start_loop())
+    await client.get_charge_points()
+
+    await client.wait_for_response()
+    hass.data[DOMAIN][config_entry.entry_id] = connector
+    hass.config_entries.async_setup_platforms(config_entry, PLATFORMS)
+
+    async def _async_disconnect_websocket(_: Event) -> None:
+        await connector.disconnect()
+
+    config_entry.async_on_unload(
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, _async_disconnect_websocket
+        )
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload the BlueCurrent config entry."""
+    connector: Connector = hass.data[DOMAIN].pop(config_entry.entry_id)
+    hass.async_create_task(connector.disconnect())
+
+    return await hass.config_entries.async_unload_platforms(config_entry, PLATFORMS)
+
+
+def set_entities_unavalible(hass: HomeAssistant, config_id: str) -> None:
+    """Set all Blue Current entities to unavailable."""
+    registry = entity_registry.async_get(hass)
+    entries = entity_registry.async_entries_for_config_entry(registry, config_id)
+
+    for entry in entries:
+        entry.write_unavailable_state(hass)
+
+
+class Connector:
+    """Define a class that connects to the Blue Current websocket API."""
+
+    charge_points: dict[str, dict] = {}
+    grid: dict[str, Any] = {}
+
+    def __init__(
+        self, hass: HomeAssistant, config: ConfigEntry, client: Client
+    ) -> None:
+        """Initialize."""
+        self._config: ConfigEntry = config
+        self._hass: HomeAssistant = hass
+        self.client: Client = client
+
+    async def connect(self, token: str) -> None:
+        """Register on_data and connect to the websocket."""
+        await self.client.connect(token)
+
+    async def on_data(self, message: dict) -> None:
+        """Handle received data."""
+
+        async def handle_charge_points(data: list) -> None:
+            """Loop over the charge points and get their data."""
+            for entry in data:
+                evse_id = entry[EVSE_ID]
+                model = entry[MODEL_TYPE]
+                self.add_charge_point(evse_id, model)
+                await self.get_charge_point_data(evse_id)
+            await self.client.get_grid_status(data[0][EVSE_ID])
+
+        object_name: str = message[OBJECT]
+
+        if DATA in message:
+            data: dict | list = message[DATA]
```

Comment:
This will leave `data` possibly unassigned in this function. I don't think that's what you want...?

Reply:
Removed the if statement.